### PR TITLE
Add several scoreboard type wrappers

### LINKED ISSUE #2617: Another 1.20.2 Issue
- [x] This issue is not solved in a development build (Using 5.1.1-Snapshot)

**Describe the bug**
I am not sure if this is a bug or I am just very stupid. 
I am creating a packetwrapper for SCOREBOARD_DISPLAY_OBJECTIVE and while trying to set the position (list, sidebar, belowName)
I get the following error 

com.comphenix.protocol.reflect.FieldAccessException: Field index 0 is out of bounds for length 0
at com.comphenix.protocol.reflect.FieldAccessException.fromFormat(FieldAccessException.java:49) ~[ProtocolLib.jar:?]
at com.comphenix.protocol.reflect.StructureModifier.write(StructureModifier.java:316) ~[ProtocolLib.jar:?]
at me.wasn325.coreapi.core.packets.WrapperPlayServerScoreboardDisplayObjective.setPosition(WrapperPlayServerScoreboardDisplayObjective.java:51) ~[coreapi.jar:?]

**To Reproduce**

https://github.com/dmulloy2/PacketWrapper/blob/master/PacketWrapper/src/main/java/com/comphenix/packetwrapper/WrapperPlayServerScoreboardDisplayObjective.java#L55 
This wrapper has almost the same structure as mine and should work for demonstrating (the problematic line is already marked)

**Expected behavior**

**Screenshots**
I am not sure if this is possible, as the problem occurs before the scoreboard is finish and is visible to the Players

**Version Info**
ProtocolLib Dump
Timestamp: 11/07/23 18:36:08

ProtocolLib Version: ProtocolLib v5.1.1-SNAPSHOT-669
Bukkit Version: Unknown-Version
Server Version: git-Purpur-"f24e535" (MC: 1.20.2)
Java Version: 17.0.8

ProtocolLib: com.comphenix.protocol.ProtocolLib@3f8ffeec[
  statistics=com.comphenix.protocol.metrics.Statistics@5cc602f1
  packetTask=com.comphenix.protocol.scheduler.DefaultTask@5463c22c
  tickCounter=1409118
  configExpectedMod=1
  updater=com.comphenix.protocol.updater.SpigotUpdater@2e37f40b
  redirectHandler=com.comphenix.protocol.ProtocolLib$2@6e0f07bf
  scheduler=com.comphenix.protocol.scheduler.DefaultScheduler@2324105
  commandProtocol=com.comphenix.protocol.CommandProtocol@480a91cd
  commandPacket=com.comphenix.protocol.CommandPacket@598ca36b
  commandFilter=com.comphenix.protocol.CommandFilter@4f28651a
  packetLogging=com.comphenix.protocol.PacketLogging@5f822f93
  skipDisable=false
  isEnabled=true
  loader=io.papermc.paper.plugin.manager.DummyBukkitPluginLoader@431a5957
  server=CraftServer{serverName=Purpur,serverVersion=git-Purpur-"f24e535",minecraftVersion=1.20.2}
  file=plugins/ProtocolLib.jar
  description=org.bukkit.plugin.PluginDescriptionFile@7fb13da4
  pluginMeta=org.bukkit.plugin.PluginDescriptionFile@7fb13da4
  dataFolder=plugins/ProtocolLib
  classLoader=PluginClassLoader{plugin=ProtocolLib v5.1.1-SNAPSHOT-669, pluginEnabled=true, url=plugins/ProtocolLib.jar}
  naggable=true
  newConfig=YamlConfiguration[path='', root='YamlConfiguration']
  configFile=plugins/ProtocolLib/config.yml
  logger=com.destroystokyo.paper.utils.PaperPluginLogger@3d8dd0be
]
Manager: com.comphenix.protocol.injector.PacketFilterManager@2cccf6b9[
  plugin=ProtocolLib v5.1.1-SNAPSHOT-669
  server=CraftServer{serverName=Purpur,serverVersion=git-Purpur-"f24e535",minecraftVersion=1.20.2}
  reporter=com.comphenix.protocol.ProtocolLib$1@5a66b299
  minecraftVersion=(MC: 1.20.2)
  asyncFilterManager=com.comphenix.protocol.async.AsyncFilterManager@457bc447
  pluginVerifier=com.comphenix.protocol.injector.PluginVerifier@16a0f476
  inboundListeners=com.comphenix.protocol.injector.SortedPacketListenerList@5d92b125
  outboundListeners=com.comphenix.protocol.injector.SortedPacketListenerList@328f3274
  registeredListeners=[PacketAdapter[plugin=CoreAPI, sending=ListeningWhitelist[priority=NORMAL, packets=[PLAYER_INFO[class=ClientboundPlayerInfoUpdatePacket, id=60]], gamephase=PLAYING, options=[]], receiving=EMPTY_WHITELIST], PacketAdapter[plugin=CoreAPI, sending=ListeningWhitelist[priority=NORMAL, packets=[SCOREBOARD_TEAM[class=PacketPlayOutScoreboardTeam, id=92]], gamephase=PLAYING, options=[]], receiving=EMPTY_WHITELIST], PacketAdapter[plugin=CoreAPI, sending=ListeningWhitelist[priority=NORMAL, packets=[NAMED_ENTITY_SPAWN[PLAY, SERVER, 235, classNames: [net.minecraft.network.protocol.game.PacketPlayOutNamedEntitySpawn, net.minecraft.network.protocol.game.ClientboundNamedEntitySpawnPacket, net.minecraft.network.play.server.SPacketSpawnPlayer] (unregistered)]], gamephase=PLAYING, options=[]], receiving=EMPTY_WHITELIST], PacketAdapter[plugin=CoreAPI, sending=EMPTY_WHITELIST, receiving=ListeningWhitelist[priority=NORMAL, packets=[TAB_COMPLETE[class=PacketPlayInTabComplete, id=10]], gamephase=PLAYING, options=[]]]]
  packetInjector=com.comphenix.protocol.injector.netty.manager.NetworkManagerPacketInjector@29cd819e
  playerInjectionHandler=com.comphenix.protocol.injector.netty.manager.NetworkManagerPlayerInjector@38dfd395
  networkManagerInjector=com.comphenix.protocol.injector.netty.manager.NetworkManagerInjector@700c4632
  debug=false
  closed=false
  injected=true
]

### PR DESCRIPTION
**Pull request by vytskalt**
Added wrappers for these NMS types (Mojang mapped names):
- [DisplaySlot](https://mappings.cephx.dev/1.20.4/net/minecraft/world/scores/DisplaySlot.html) (added in 1.20.2)
- [ObjectiveCriteria$RenderType](https://mappings.cephx.dev/1.20.4/net/minecraft/world/scores/criteria/ObjectiveCriteria$RenderType.html)
- [ChatFormatting](https://mappings.cephx.dev/1.20.4/net/minecraft/ChatFormatting.html)
- [ClientboundSetPlayerTeamPacket$Parameters](https://mappings.cephx.dev/1.20.4/net/minecraft/network/protocol/game/ClientboundSetPlayerTeamPacket$Parameters.html)
- [Style](https://mappings.cephx.dev/1.20.4/net/minecraft/network/chat/Style.html)
- [NumberFormat](https://mappings.cephx.dev/1.20.4/net/minecraft/network/chat/numbers/NumberFormat.html) (sum type)

I also fixed a compile error. ~~Marked as draft for now as I haven't fully tested it yet.~~

Closes #2617

### CODE REVIEW

**Comment by vytskalt** [comment]
I have manually tested everything with the hacky code below on 1.8.9, 1.12.2, 1.13, 1.16.5, 1.17, 1.20.2 and 1.20.4. Everything appears to be working correctly.

```java
public void testProtocolLib() {
  ProtocolManager pm = ProtocolLibrary.getProtocolManager();
  if (new MinecraftVersion(1, 20, 2).atOrAbove()) {
    getLogger().info("Testing DisplaySlot (1.20.2+)");
    pm.createPacket(PacketType.Play.Server.SCOREBOARD_DISPLAY_OBJECTIVE)
      .getDisplaySlots()
      .write(0, EnumWrappers.DisplaySlot.LIST);
  }

  // expect parameters to be supported on 1.17+
  Preconditions.checkState(WrappedTeamParameters.isSupported() == MinecraftVersion.CAVES_CLIFFS_1.atOrAbove());

  if (MinecraftVersion.CAVES_CLIFFS_1.atOrAbove()) {
    getLogger().info("Testing WrappedTeamParameters (1.17+)");
    WrappedChatComponent c = WrappedChatComponent.fromText("test");
    WrappedTeamParameters p = WrappedTeamParameters.newBuilder()
      .displayName(c)
      .prefix(c)
      .suffix(c)
      .nametagVisibility("never")
      .collisionRule("always")
      .color(EnumWrappers.ChatFormatting.OBFUSCATED)
      .options(1)
      .build();

    Preconditions.checkState(p.getDisplayName().getHandle() == c.getHandle());
    Preconditions.checkState(p.getPrefix().getHandle() == c.getHandle());
    Preconditions.checkState(p.getSuffix().getHandle() == c.getHandle());
    Preconditions.checkState(p.getNametagVisibility().equals("never"));
    Preconditions.checkState(p.getCollisionRule().equals("always"));
    Preconditions.checkState(p.getColor() == EnumWrappers.ChatFormatting.OBFUSCATED);
    Preconditions.checkState(p.getOptions() == 1);
  }

  getLogger().info("Testing ChatFormatting");
  Object colorHandle = EnumWrappers.getChatFormattingConverter().getGeneric(EnumWrappers.ChatFormatting.DARK_AQUA);
  Preconditions.checkState(EnumWrappers.ChatFormatting.DARK_AQUA == EnumWrappers.getChatFormattingConverter().getSpecific(colorHandle));

  Preconditions.checkState(WrappedNumberFormat.isSupported() == MinecraftVersion.v1_20_4.atOrAbove());

  getLogger().info("Testing RenderType");
  pm.createPacket(PacketType.Play.Server.SCOREBOARD_OBJECTIVE)
    .getRenderTypes()
    .write(0, EnumWrappers.RenderType.HEARTS);

  {
    getLogger().info("Testing WrappedComponentStyle");
    JsonObject style = (JsonObject) new JsonParser().parse("{\"color\":\"green\",\"bold\":true}");
    WrappedComponentStyle wrappedStyle = WrappedComponentStyle.fromJson(style);
    Preconditions.checkState(wrappedStyle.getJson().equals(style));
  }

  getLogger().info("All tests passed!");
}
```

**Comment by vytskalt** [comment]
Also FYI @dmulloy2 the tests don't currently run because of a version mismatch between spigot-api and spigot. org.spigotmc:spigot:1.20.4-R0.1-SNAPSHOT in the https://repo.dmulloy2.net/ repo needs to be updated to the latest version.

**Comment by vytskalt** [comment]
@dmulloy2 pls let me know if you are interested in this PR so I can know if I should update it for 1.20.6

**Comment by dmulloy2** [comment]
@vytskalt this looks great! Would love if you could update it for 1.20.6. Seems like it's failing to compile because those new methods already exist in serialized offline player

File: src/main/java/com/comphenix/protocol/events/SerializedOfflinePlayer.java [hunk 1]
```diff
@@ -291,6 +291,18 @@ public void setWhitelisted(boolean whitelisted) {
         this.whitelisted = whitelisted;
     }
 
+    @Nullable
+    @Override
+    public Location getRespawnLocation() {
```
Comment (dmulloy2) [inline src/main/java/com/comphenix/protocol/events/SerializedOfflinePlayer.java:296]: no longer necessary

**Review by dmulloy2 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**